gadget/update.go

```diff
@@ -424,6 +425,161 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
+			Size:               quantity.Size(part.SizeInBytes),
+			Offset:             quantity.Offset(part.StartInBytes),
+			PartitionUUID:      part.PartitionUUID,
+			OriginalKernelPath: part.KernelDeviceNode,
+			OriginalDevicePath: part.KernelDevicePath,
+			PartitionType:      part.PartitionType,
+			PartitionLabel:     part.PartitionLabel,  // this will be empty on dos disks
+			FilesystemLabel:    part.FilesystemLabel, // blkid encoded
+			FilesystemUUID:     part.FilesystemUUID,  // blkid encoded
+			FilesystemType:     part.FilesystemType,
+		}
+
+		mappedStructures = append(mappedStructures, ms)
+
+		// delete this partition from the map
+		delete(diskPartitionsByOffset, uint64(structure.StartOffset))
+	}
+
+	// We should have deleted all structures from diskPartitionsByOffset that
+	// are in the gadget.yaml laid out volume, however there is a small
+	// possibility (mainly due to bugs) where we could still have partitions in
+	// diskPartitionsByOffset. So we check to make sure there are no partitions
+	// left over.
+	// However, the one notable exception to this is in the case of legacy UC16
+	// or UC18 gadgets where the system-data role could have been left out and
+	// ubuntu-image would dynamically create the partition. In this case, we
+	// ought to just ignore this on-disk structure since it is not in the
+	// gadget.yaml, and the primary use case of tracking disks and structures is
+	// for gadget asset update, but by definition something which is not in the
+	// gadget.yaml cannot be updated via gadget asset updates.
+	switch len(diskPartitionsByOffset) {
+	case 0:
+		// expected, no implicit system-data
+		break
+	case 1:
+		// could be implicit system-data
+		if opts.AllowImplicitSystemData {
+			var part disks.Partition
+			for _, part = range diskPartitionsByOffset {
+				break
+			}
+
+			s, err := OnDiskStructureFromPartition(part)
+			if err != nil {
+				return res, err
+			}
+
+			if onDiskStructureIsLikelyImplicitSystemDataRole(expLayout, diskLayout, s) {
+				// it is likely the implicit system-data
+				logger.Debugf("Identified implicit system-data role on system as %s", s.Node)
+				break
+			}
+		}
+		fallthrough
+	default:
+		// we for sure have left over partitions that should have been in the
+		// gadget.yaml - make a nice string with what partitions are leftover
+		leftovers := []string{}
+		for _, part := range diskPartitionsByOffset {
+			leftovers = append(leftovers, part.KernelDeviceNode)
+		}
+		// this is an internal error because to get here we would have had to
+		// pass validation in EnsureLayoutCompatibility but then still have
+		// extra partitions - the only non-buggy situation where that function
+		// passes validation but leaves partitions on disk not in the YAML is
+		// the implicit system-data role handled above
+		return res, fmt.Errorf("internal error: unexpected additional partitions on disk %s not present in the gadget layout: %v", disk.KernelDeviceNode(), leftovers)
```

Comment:
I guess it's not that easy to procure this in the tests without intercepting EnsureLayoutCompatibility?

Reply:
yes we would need to mock EnsureLayoutCompatibility to essentially act buggy to get here